### PR TITLE
feat(replays): Add is_archived column and migration

### PR DESCRIPTION
**Pull request by cmanallen**
- Adds new is_archived column.

### CODE REVIEW

File: snuba/snuba_migrations/replays/0006_add_is_archived_column.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+new_columns: Sequence[Tuple[Column[Modifiers], str]] = [
+    (Column("is_archived", UInt(8, Modifiers(nullable=True))), "urls")
+]
+
+alter_forward: Sequence[Column[Modifiers]] = [
+    Column("platform", String(Modifiers(nullable=True)))
+]
+
+alter_backward: Sequence[Column[Modifiers]] = [
+    Column("platform", String(Modifiers(nullable=False, default="''")))
```
Comment (evanh) [inline snuba/snuba_migrations/replays/0006_add_is_archived_column.py:17]: Is the new column supposed to have a default? Have you confirmed that running the `MODIFY` with just the nullable removes the default = ''?

**Review by evanh — COMMENTED** [review]
Adding the column looks fine to me, I am curious about the change to platform though.

Why is platform now nullable? I am not sure of the impact of running that `MODIFY`, Clickhouse might lock up while that runs.

**Comment by onewland** [comment]
Is it expected that the state of `is_archived` will change for a record over time? If yes, have you discussed the implications of mutability in a table with somebody in SNS?

Also do you need to be able to distinguish between empty string and null for platform? Because that's all that the nullable modifier buys you.

**Comment by cmanallen** [comment]
@onewland 

> Is it expected that the state of `is_archived` will change for a record over time? If yes, have you discussed the implications of mutability in a table with somebody in SNS?

Replays are not mutable.  A Replay is an aggregation of many rows.  Replays emulate mutability in that new rows influence the aggregated output.

> Also do you need to be able to distinguish between empty string and null for `platform`? Because that's all that the nullable modifier in the migration buys you. See https://clickhouse.com/docs/en/sql-reference/data-types/nullable/#storage-features

Nullability provides benefits to aggregate functions like `groupUniqArray` which will filter null values automatically.  Otherwise we have to `map` in the `select` or in the application server.

The Replays schema has 19 nullable fields.

File: snuba/snuba_migrations/replays/0006_add_is_archived_column.py (hunk 1 above)
Comment (cmanallen) [reply snuba/snuba_migrations/replays/0006_add_is_archived_column.py:17]: @evanh If this is going to cause operational issues we can drop the nullability requirements for platform.

Comment (evanh) [reply snuba/snuba_migrations/replays/0006_add_is_archived_column.py:17]: It's not an operational concern, it's just that I'm not sure if the `default = ''` is preserved in your migration. If it is, you'll have a nullable column that is always setting a default of `''`. The default should appear when you do `DESCRIBE TABLE`. To test, I would see that it is there before the migration, run the migration, and then be sure it's gone.

Comment (cmanallen) [reply snuba/snuba_migrations/replays/0006_add_is_archived_column.py:17]: I'm going to drop the nullability requirements for platform.  If the need arises in the future, we can address it then.

Re: migration misunderstandings.  I'm looking at the SQL output and this ALTER statement makes sense to me.  Where am I going wrong?  If I'm altering a nullable column to not-null column it needs a default on the column, right?

> ALTER TABLE replays_local MODIFY COLUMN platform String DEFAULT '';

Comment (evanh) [reply snuba/snuba_migrations/replays/0006_add_is_archived_column.py]: Ah I see, yes OK this is fine. Thank you.

**Review by evanh — APPROVED** [review]

**Pull request merged**